File: 004/solution.rb
def solution(factor_range)
  factor_range.reduce(0) do |result, x|
    factor_range.reduce(result) do |result, y|
      product = x * y
      product > result && palindrome?(product) ? product : result
    end
  end
end

def palindrome?(object)
  object.to_s == object.to_s.reverse
end

if __FILE__ == $0
  puts solution(100...1000)
end

File: 004/solution.py
def solution(factor_range):
  result = 0
  for x in factor_range:
    for y in factor_range:
      product = x * y
      if product > result and palindrome(product):
        result = product
  return result

def palindrome(number):
  return str(number) == str(number)[::-1]

if __name__ == '__main__':
  print solution(xrange(100, 1000))
